File: ios/BiblosSurfer/Views/Reader/ReaderViewModel.swift
//
//  ReaderViewModel.swift
//  BiblosSurfer
//
//  Created by Mikołaj Linczewski on 21/08/2026.
//

import Foundation
import Observation
import ReadiumNavigator
import ReadiumShared
import UIKit

protocol ReaderActions: AnyObject {
    func playPauseTTS()
    func stopTTS()
    func nextUtterance()
    func previousUtterance()
    func applyReaderSettings()
    func closeSettings()
    func dismissPresentedError()
}

struct ReaderViewState: Observable {
    var title: String = ""
    var isPlaying = false
    var canSpeak = false
    var format: PublicationFormat = .unknown
    var error: DescriptiveError?
}

@Observable
final class ReaderViewModel: ErrorDismissing {
    var viewProperties: ReaderViewState
    let settings: ReaderSettingsStore
    weak var actions: ReaderActions?
    var availableVoices: [TTSVoice] = []
    var settingsEpoch = 0

    init(title: String, format: PublicationFormat, canSpeak: Bool, settings: ReaderSettingsStore = ReaderSettingsStore()) {
        self.settings = settings
        self.viewProperties = ReaderViewState(
            title: title,
            canSpeak: canSpeak && format == .epub,
            format: format
        )
    }

    func apply(ttsState: PublicationSpeechSynthesizer.State) {
        switch ttsState {
        case .stopped:
            viewProperties.isPlaying = false
        case .paused:
            viewProperties.isPlaying = false
        case .playing:
            viewProperties.isPlaying = true
        }
    }

    func presentError(_ error: DescriptiveError) {
        viewProperties.error = error
    }

    func playPause() {
        actions?.playPauseTTS()
    }

    func stopReading() {
        actions?.stopTTS()
        
    }

    func nextUtterance() {
        actions?.nextUtterance()
    }

    func previousUtterance() {
        actions?.previousUtterance()
    }

    var ttsLanguages: [Language] {
        _ = settingsEpoch
        var seen = Set<String>()
        var languages: [Language] = []
        for voice in availableVoices {
            let language = voice.language.removingRegion()
            if seen.insert(language.code.bcp47).inserted {
                languages.append(language)
            }
        }
        return languages.sorted { $0.localizedDescription() < $1.localizedDescription() }
    }

    var ttsVoices: [TTSVoice] {
        _ = settingsEpoch
        guard let code = settings.defaultLanguage else {
            return availableVoices.rankedByAppleQuality()
        }
        return availableVoices
            .filterByLanguage(Language(code: .bcp47(code)))
            .rankedByAppleQuality()
    }

    func selectTTSLanguage(_ code: String?) {
        settings.defaultLanguage = code
        if let code, let voiceId = settings.voiceIdentifier,
           let voice = availableVoices.first(where: { $0.identifier == voiceId }),
           voice.language.removingRegion().code.bcp47 != Language(code: .bcp47(code)).removingRegion().code.bcp47 {
            settings.voiceIdentifier = nil
        }
        settingsDidChange()
    }

    func settingsDidChange() {
        settingsEpoch += 1
        actions?.applyReaderSettings()
    }

    func closeSettings() {
        actions?.closeSettings()
    }

    func dismissError() {
        viewProperties.error = nil
        actions?.dismissPresentedError()
    }
}

File: android/app/src/main/java/miko/biblossurfer/ui/reader/ReaderViewModel.kt
package miko.biblossurfer.ui.reader

import miko.biblossurfer.data.DescriptiveError
import miko.biblossurfer.data.ReaderSettingsStore
import miko.biblossurfer.data.model.PublicationFormat
import miko.biblossurfer.data.tts.TtsPlaybackState
import miko.biblossurfer.data.tts.TtsVoiceInfo
import miko.biblossurfer.data.tts.filterByLanguage
import miko.biblossurfer.data.tts.rankedByAppleQuality
import miko.biblossurfer.ui.library.ErrorDismissing
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

interface ReaderActions {
    fun playPauseTTS()
    fun stopTTS()
    fun nextUtterance()
    fun previousUtterance()
    fun applyReaderSettings()
    fun closeSettings()
    fun dismissPresentedError()
}

data class ReaderViewState(
    val title: String = "",
    val isPlaying: Boolean = false,
    val canSpeak: Boolean = false,
    val format: PublicationFormat = PublicationFormat.UNKNOWN,
    val error: DescriptiveError? = null,
    val showSettings: Boolean = false,
    val showTtsPanel: Boolean = false,
)

class ReaderViewModel(
    title: String,
    format: PublicationFormat,
    canSpeak: Boolean,
    val settings: ReaderSettingsStore,
) : ErrorDismissing {
    var actions: ReaderActions? = null
    var availableVoices: List<TtsVoiceInfo> = emptyList()
    var settingsEpoch: Int = 0
        private set

    private val _viewProperties = MutableStateFlow(
        ReaderViewState(
            title = title,
            canSpeak = canSpeak && format == PublicationFormat.EPUB,
            format = format,
        )
    )
    val viewProperties: StateFlow<ReaderViewState> = _viewProperties.asStateFlow()

    fun apply(ttsState: TtsPlaybackState) {
        _viewProperties.update { state ->
            when (ttsState) {
                TtsPlaybackState.STOPPED -> state.copy(isPlaying = false, showTtsPanel = false)
                TtsPlaybackState.PAUSED -> state.copy(isPlaying = false, showTtsPanel = true)
                TtsPlaybackState.PLAYING -> state.copy(isPlaying = true, showTtsPanel = true)
            }
        }
    }

    fun presentError(error: DescriptiveError) {
        _viewProperties.update { it.copy(error = error) }
    }

    fun playPause() {
        actions?.playPauseTTS()
    }

    fun stopReading() {
        actions?.stopTTS()
    }

    fun nextUtterance() {
        actions?.nextUtterance()
    }

    fun previousUtterance() {
        actions?.previousUtterance()
    }

    fun ttsLanguages(): List<String> {
        val seen = linkedSetOf<String>()
        for (voice in availableVoices) {
            seen.add(voice.languageWithoutRegion)
        }
        return seen.sorted()
    }

    fun ttsVoices(): List<TtsVoiceInfo> {
        val code = settings.defaultLanguage ?: return availableVoices.rankedByAppleQuality()
        return availableVoices.filterByLanguage(code).rankedByAppleQuality()
    }

    fun selectTTSLanguage(code: String?) {
        settings.defaultLanguage = code
        if (code != null) {
            val voiceId = settings.voiceIdentifier
            val voice = availableVoices.firstOrNull { it.identifier == voiceId }
            if (voice != null && voice.languageWithoutRegion != code.lowercase().substringBefore("-")) {
                settings.voiceIdentifier = null
            }
        }
        settingsDidChange()
    }

    fun settingsDidChange() {
        settingsEpoch += 1
        actions?.applyReaderSettings()
    }

    fun openSettings() {
        _viewProperties.update { it.copy(showSettings = true) }
    }

    fun closeSettings() {
        _viewProperties.update { it.copy(showSettings = false) }
        actions?.closeSettings()
    }

    override fun dismissError() {
        _viewProperties.update { it.copy(error = null) }
        actions?.dismissPresentedError()
    }
}
